multimedia/VLC/files/patch-build-on-pre-1012.diff

```diff
@@ -73,6 +37,112 @@ index 38777226cfe56211aa0bd1efc9623ac7ca4cf650..d31f2168885b9b2d68d7dc562ea7a484
+-
+-@end
+-#endif
+-
+ /*****************************************************************************
+  * Struct
+  *****************************************************************************/
+@@ -221,6 +213,11 @@ static int Open(vlc_object_t *p_this)
+             return VLC_EGENERIC;
+         }
+ 
++// only compile this block if the requestAccessForMediaType method is supported;
++// the code inside `if (@available(foo)) {}` is only *executed* when foo is true but
++// it is always parsed at compile time; @available is not an #ifdef that protects against
++// undefined types or functions.
++#ifdef MAC_OS_X_VERSION_10_14
+         if (@available(macOS 10.14, *)) {
+             msg_Dbg(p_demux, "Check user consent for access to the audio device");
+ 
+@@ -240,6 +237,7 @@ static int Open(vlc_object_t *p_this)
+                 return VLC_EGENERIC;
+             }
+         }
++#endif
+ 
+         NSError *error = nil;
+         AVCaptureDeviceInput *input = [AVCaptureDeviceInput deviceInputWithDevice:device error:&error];
+diff --git a/modules/audio_output/coreaudio_common.c b/modules/audio_output/coreaudio_common.c
+index 69fd47137050a1d21ac7eadfbfaafe6f557b8ee9..4a58744104a945d1d11014f9b9948f3dd28dde34 100644
+--- a/modules/audio_output/coreaudio_common.c
++++ b/modules/audio_output/coreaudio_common.c
+@@ -25,6 +25,28 @@
+ #import "coreaudio_common.h"
+ #import <CoreAudio/CoreAudioTypes.h>
+ 
++#if __MAC_OS_X_VERSION_MAX_ALLOWED < 101200
++#include <dlfcn.h>
++static struct
++{
++    void (*lock)(os_unfair_lock *lock);
++    void (*unlock)(os_unfair_lock *lock);
++} unfair_lock;
++
++static void
++ca_init_once(void)
++{
++    unfair_lock.lock = dlsym(RTLD_DEFAULT, "os_unfair_lock_lock");
++    if (!unfair_lock.lock)
++        return;
++    unfair_lock.unlock = dlsym(RTLD_DEFAULT, "os_unfair_lock_unlock");
++    if (!unfair_lock.unlock)
++        unfair_lock.lock = NULL;
++}
++#define os_unfair_lock_lock unfair_lock.lock
++#define os_unfair_lock_unlock unfair_lock.unlock
++#endif
++
+ static inline uint64_t
+ BytesToFrames(struct aout_sys_common *p_sys, size_t i_bytes)
+ {
+@@ -114,6 +136,10 @@ ca_Open(audio_output_t *p_aout)
+ 
+     if (mach_timebase_info(&p_sys->tinfo) != KERN_SUCCESS)
+         return VLC_EGENERIC;
++#if __MAC_OS_X_VERSION_MAX_ALLOWED < 101200
++    static pthread_once_t once = PTHREAD_ONCE_INIT;
++    pthread_once(&once, ca_init_once);
++#endif
+ 
+     assert(p_sys->tinfo.denom != 0 && p_sys->tinfo.numer != 0);
+ 
+diff --git a/modules/video_output/caopengllayer.m b/modules/video_output/caopengllayer.m
+index bff225525d9eb68dace1d5839857fd61e3fd83cf..ce11bdfbc182632592c5e0ba19058187d22b8d20 100644
+--- a/modules/video_output/caopengllayer.m
++++ b/modules/video_output/caopengllayer.m
+@@ -40,6 +40,18 @@
+ #include <vlc_atomic.h>
+ 
+ #import <QuartzCore/QuartzCore.h>
++
++// no idea really what version check is needed here
++#if __MAC_OS_X_VERSION_MAX_ALLOWED < 101100
```

Comment:
Well how did you decide that a version check was needed?

Reply:
Because the build fails on 10.9 without that check.

Reply:
What is the error message on 10.9 without that check?

Reply:
Just curiosity I presume? O:-)

As you noticed, the version check protects some additional code, a protocol definition that's missing on 10.9 .

```
video_output/caopengllayer.m:106:40: error: cannot find protocol declaration for 'CALayerDelegate'; did you mean 'NSDrawerDelegate'?
@interface VLCVideoLayerView : NSView <CALayerDelegate>
                                       ^~~~~~~~~~~~~~~
                                       NSDrawerDelegate
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.10.sdk/System/Library/Frameworks/AppKit.framework/Headers/NSDrawer.h:89:11: note: 'NSDrawerDelegate' declared here
@protocol NSDrawerDelegate <NSObject>
          ^
1 error generated.
```

I haven't been able to find out when that protocol was introduced.

Reply:
> Just curiosity I presume? O:-)

I'm trying to help you answer the question you didn't know the answer to: what version of macOS introduced this support.

What I'm finding is that CALayerDelegate was an "informal protocol" for awhile before it was [formalized in 10.12](https://developer.apple.com/documentation/quartzcore/calayerdelegate). I don't know exactly what an informal protocol is, but I found [this code in chromium](https://chromium.googlesource.com/chromium/src/+/b7413b8f60c48c11d36bc15023bc2c9d08c15406%5E%21/#F0):

```
#if !defined(MAC_OS_X_VERSION_10_12) || \
    MAC_OS_X_VERSION_MAX_ALLOWED < MAC_OS_X_VERSION_10_12

// The protocol was formalized by the 10.12 SDK, but it was informally used
// before.
@protocol CAAnimationDelegate
- (void)animationDidStart:(CAAnimation*)animation;
- (void)animationDidStop:(CAAnimation*)animation finished:(BOOL)finished;
@end

@protocol CALayerDelegate
@end

#endif  // MAC_OS_X_VERSION_10_12
```

Maybe code like that helps you make VLC compatible with older systems.

Reply:
I was guessing you were, appreciated.

An informal protocol could be one that responds to a need and finds itself being used (and implemented) by more and more projects. IIRC an ObjC protocol is a way to extend a class without subclassing, something anyone can do - the question would be why they formalised it as a protocol.

I wouldn't have thought to look for answers in the Chromium code - do I understand correctly that my check should include 10.11 then?

Reply:
Yes my guess is it should be `< 101200`.

I didn't seek out Chromium's code specifically; that's just what turned up when I Googled CALayerDelegate.